bellman/src/domain/linear.rs

```diff
@@ -0,0 +1,163 @@
+use std::ops::{Add, Sub};
+use ff::{ScalarEngine, Field};
+
+/// Represents a variable in our constraint system.
+#[derive(Copy, Clone, Debug)]
+pub struct Variable(Index);
+
+impl Variable {
+    /// This constructs a variable with an arbitrary index.
+    /// Circuit implementations are not recommended to use this.
+    pub fn new_unchecked(idx: Index) -> Variable {
+        Variable(idx)
+    }
+
+    /// This returns the index underlying the variable.
+    /// Circuit implementations are not recommended to use this.
+    pub fn get_unchecked(&self) -> Index {
+        self.0
+    }
+}
+
+/// Represents the index of either an input variable or
+/// auxiliary variable.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum Index {
+    Input(usize),
+    Aux(usize),
+}
+
+/// This represents a linear combination of some variables, with coefficients
+/// in the scalar field of a pairing-friendly elliptic curve group.
+#[derive(Clone)]
+pub struct LinearCombination<E: ScalarEngine>(pub Vec<(Variable, E::Fr)>);
+
+impl<E> AsRef<[(Variable, E::Fr)]> for LinearCombination<E> 
+where
+    E: ScalarEngine
+{
+    fn as_ref(&self) -> &[(Variable, E::Fr)] {
+        &self.0
+    }
+}
+
+impl<E> LinearCombination<E> 
+where
+    E: ScalarEngine
+{
+    pub fn zero() -> Self {
+        LinearCombination(vec![])
+    }
+}
+
+impl<E> Add<(E::Fr, Variable)> for LinearCombination<E> 
+where
+    E: ScalarEngine
+{
+    type Output = Self;
+
+    fn add(mut self, (coeff, var): (E::Fr, Variable)) -> LinearCombination<E> {
+        self.0.push((var, coeff));
+
+        self
```

Comment:
Is this an implicit return? Looks cool, would be keen to learn more about it.

Reply:
Yeah last line in rust without a semicolon will be a return

Reply:
Awesome